crates/ra_hir/src/source_binder.rs

```diff
@@ -409,6 +410,45 @@ impl SourceAnalyzer {
         crate::ty::autoderef(db, &self.resolver, canonical).map(|canonical| canonical.value)
     }
 
+    /// Checks that particular type `ty` implements `std::future::Future` trait.
+    /// This function is used in `.await` syntax completion.
+    pub fn impls_future(&self, db: &impl HirDatabase, ty: Ty) -> bool {
+        // Search for std::future::Future trait in scope
```

Comment:
Please search for the `future_trait` lang item instead using for example `db.lang_item(krate, "future_trait".into())`. That makes it work for no_std.

https://github.com/rust-lang/rust/blob/1df512fcaeaf17639c5d28a3045814d6f7a7db97/src/libcore/future/future.rs#L30

Reply:
Hmm. rustc uses the path in the actual lowering though (it just has special-casing for no_std), the lang item seems to only be used for diagnostics. So I'm not sure relying on the lang item is a good idea...

If you do use a path, please resolve it properly instead of relying on traits_in_scope though. [Here](https://github.com/rust-analyzer/rust-analyzer/blob/831d39b0f7ed0041cbc2fb8beb4cb375ab9652b1/crates/ra_hir/src/ty/infer.rs#L1399) is the existing place where the `Future::Output` associated type is resolved. It would be nice to extract these well-known paths into their own module.

Reply:
@flodiebold can I try just both paths `std::futures::Future` and `core::futures::Future` instead of relying on lang item?

Reply:
@flodiebold can you please suggest how to mock `Future` trait in tests? When I do
```rust
            pub mod std {
                pub mod future {
                    pub trait Future {}
                }
            }
```

for some reason, resolution by the absolute path with following code results in None:
```rust
        let std_future_path = Path {
            kind: PathKind::Abs,
            segments: vec![
                PathSegment { name: name::STD, args_and_bindings: None },
                PathSegment { name: name::FUTURE_MOD, args_and_bindings: None },
                PathSegment { name: name::FUTURE_TYPE, args_and_bindings: None },
            ],
        };

        dbg!(self.resolver.resolve_path_segments(db, &std_future_path).into_fully_resolved());

        let std_future_trait =
            match self.resolver.resolve_path_segments(db, &std_future_path).into_fully_resolved() {
                PerNs { types: Some(Resolution::Def(ModuleDef::Trait(trait_))), .. } => {
                    Some(trait_)
                }
                _ => None,
            };
```

Reply:
> @flodiebold can I try just both paths `std::futures::Future` and `core::futures::Future` instead of relying on lang item?

We don't handle `no_std` specially yet, so there's no sense in doing that currently. I think it'd be enough to have all well-known paths in one place, so that when we add support for `no_std`, we can do the special-casing there.

> @flodiebold can you please suggest how to mock `Future` trait in tests?

`std` needs to be a separate crate, like [here](https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ra_hir/src/ty/tests.rs#L24). The problem is that you can't specify the crate graph with the completion test functions, but I think you can do it this way:

```rust
//- /lib.rs
use std::future::*;
struct A {}
impl Future for A {}
fn foo(a: A) {
    a.<|>
}

//- /std/lib.rs
mod futures {
    trait Future {}
}
```
(because of [this code](https://github.com/rust-analyzer/rust-analyzer/blob/0ca30b6c4bb717e3cc54bfc98ffaa7a6d8c2cdfd/crates/ra_ide_api/src/mock_analysis.rs#L96))